File: sparktk-core/src/main/scala/org/trustedanalytics/sparktk/frame/internal/ops/matrix/MatrixSvd.scala
/**
 *  Copyright (c) 2016 Intel Corporation 
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.trustedanalytics.sparktk.frame.internal.ops.matrix

import org.apache.spark.rdd.RDD
import org.apache.spark.sql.catalyst.expressions.GenericRow
import org.apache.spark.sql.Row
import org.apache.spark.mllib.linalg.{ DenseMatrix => DM, Matrix, Matrices }
import breeze.linalg.{ DenseMatrix => BDM, Matrix => BM }
import org.trustedanalytics.sparktk.frame.{ Frame, Column, DataTypes }
import org.trustedanalytics.sparktk.frame.internal._
import org.trustedanalytics.sparktk.frame.internal.rdd.FrameRdd

trait MatrixSvdTransform extends BaseFrame {

  def matrixSvd(matrixColumnName: String): Unit = {
    execute(MatrixSvd(matrixColumnName))
  }
}

case class MatrixSvd(matrixColumnName: String) extends FrameTransform {

  require(matrixColumnName != null, "Matrix column name cannot be null")

  override def work(state: FrameState): FrameState = {

    val frame = new Frame(state.rdd, state.schema)

    frame.schema.requireColumnIsType(matrixColumnName, DataTypes.matrix)
    //run the operation
    frame.addColumns(MatrixSvd.matrixSvd(matrixColumnName), Seq(Column("U_" + matrixColumnName, DataTypes.matrix),
      Column("V_" + matrixColumnName, DataTypes.matrix),
      Column("SingularVectors_" + matrixColumnName, DataTypes.matrix)))
    FrameState(frame.rdd, frame.schema)

  }
}

object MatrixSvd extends Serializable {
  /**
   * Computes the singular value decomposition for each matrix of the frame
   */
  def matrixSvd(matrixColumnName: String)(rowWrapper: RowWrapper): Row = {

    val matrix = rowWrapper.value(matrixColumnName).asInstanceOf[DM]
    val breezeMatrix = MatrixFunctions.asBreeze(matrix)

    val matrixSvdResult = breeze.linalg.svd(breezeMatrix)

    val uMatrix = MatrixFunctions.fromBreeze(matrixSvdResult.U).asInstanceOf[DM]
    val vtMatrix = MatrixFunctions.fromBreeze(matrixSvdResult.Vt).asInstanceOf[DM]
    val singularVectors = new DM(1, matrixSvdResult.singularValues.length, matrixSvdResult.singularValues.toArray, false)

    Row.apply(uMatrix, vtMatrix, singularVectors)
  }

}

File: python/sparktk/frame/ops/matrix_svd.py
# vim: set encoding=utf-8

#  Copyright (c) 2016 Intel Corporation 
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

def matrix_svd(self, matrix_column_name):

    """

    Compute the Singular Value Decomposition of a matrix

    Parameters
    ----------

    :param matrix_column_name: Name of the column to compute the svd on
    :return: (Frame) returns the frame with three new columns storing the U matrix, Vt matrix and Singular Vectors

    Calculate the Singular Value Decomposition for each matrix in column 'matrix_column_name'

    Examples
    --------
        >>> from sparktk import dtypes
        >>> data = [[1, [[1,2,3,5],[2,3,5,6],[4,6,7,3],[8,9,2,4]]]]
        >>> schema = [('id', int),('pixeldata', dtypes.matrix)]
        >>> my_frame = tc.frame.create(data, schema)

        >>> my_frame.inspect()
        [#]  id  pixeldata
        ============================
        [0]   1  [[ 1.  2.  3.  5.]
        [ 2.  3.  5.  6.]
        [ 4.  6.  7.  3.]
        [ 8.  9.  2.  4.]]

        Compute the singular value decomposition for the matrices in 'pixeldata' column of the frame
        >>> my_frame.matrix_svd('pixeldata')

        Three new columns get added storing the U matrix, Vt matrix and Singular Vectors
        >>> my_frame.inspect()
        [#]  id  pixeldata
        ============================
        [0]   1  [[ 1.  2.  3.  5.]
         [ 2.  3.  5.  6.]
         [ 4.  6.  7.  3.]
         [ 8.  9.  2.  4.]]
        <BLANKLINE>
        [#]  U_pixeldata
        ========================================================
        [0]  [[-0.29128979 -0.43716238 -0.44530839  0.72507913]
         [-0.42474933 -0.55066945 -0.26749936 -0.66692972]
         [-0.55099141 -0.16785045  0.79986267  0.16868433]
         [-0.65661765  0.69099814 -0.30060644 -0.0317899 ]]
        <BLANKLINE>
        [#]  Vt_pixeldata
        ========================================================
        [0]  [[-0.47195872 -0.60780067 -0.44835972 -0.45476024]
         [ 0.50289367  0.40702574 -0.58469285 -0.48945099]
         [-0.05244699  0.11313693  0.65644993 -0.74399115]
         [-0.72222035  0.67239008 -0.16180641  0.01039344]]
        <BLANKLINE>
        [#]  SingularVectors_pixeldata
        ============================================================
        [0]  [[ 18.21704938   6.59797925   3.54086993   0.26080987]]

    """

    self._scala.matrixSvd(matrix_column_name)